### ISSUE TITLE
[FEATURE_REQUEST] Fragment search for Manage chat files

### ISSUE THREAD

**Issue by Technologicat** [SillyTavern/SillyTavern#1797]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

Fragment search (a.k.a. swoop) is incredibly efficient at finding a single line in a large collection of text. The *Timelines* extension uses it now.

It might be convenient to have it also in the *Manage chat files* view.

But since I'm not sure if that's desirable, I'm posting this instead of directly making a PR. :)

### Describe the solution you'd like

**What fragment search is**: The search query consists of whitespace-separated strings, called *fragments*. The search ANDs them, i.e. all fragments must match. The ordering of the fragments in the query doesn't matter. E.g. `frag mat que` would find this line, and this line only.

Here's a quick prototype that implements this. This can be dropped in to replace the original `const filteredData = ...` in `displayChats` in `SillyTavern/public/script.js`:

```javascript
        const filteredData = data.filter(chat => {
            const fileName = chat['file_name'];
            const chatContent = rawChats[fileName];

            // fragment search a.k.a. swoop
            function makeQueryFragments(query) {
                let fragments = query.trim().split(/\s+/).map( function (str) { return str.trim(); } );
                fragments = [...new Set(fragments)];  // uniques only
                // fragments = fragments.filter( function(str) { return str.length >= 3; } );  // Helm in Emacs does this, but perhaps better if we don't.
                fragments = fragments.map( function (str) { return str.toLowerCase(); } );
                return fragments;
            }
            function matchFragments(fragments, text) {
                if (!text) {
                    return false;
                }
                // https://github.com/Technologicat/js-for-pythonistas
                function all (predicate, iterable) {  // Same semantics as in Python.
                    for (const x of iterable) {
                        if (!predicate(x)) {
                            return false;
                        }
                    }
                    return true;
                }
                return all(function (str) { return text.includes(str); }, fragments);
            }
            const fragments = makeQueryFragments(searchQuery);
            return chatContent && Object.values(chatContent).some(message => matchFragments(fragments, message?.mes?.toLowerCase()));

            // // classical substring search
            // return chatContent && Object.values(chatContent).some(message => message?.mes?.toLowerCase()?.includes(searchQuery.toLowerCase()));
        });
```

Speaking as a pythonista. With some proper JavaScript-fu, this could probably be made shorter.

The speed seems decent at least in my very informal tests.

### Describe alternatives you've considered

- Having fragment search as an toggleable option, in advanced user settings.
  - But in my opinion, this should default to on:
    - Searching with just one fragment is the same as a classical substring search. The only use case this breaks are search strings containing whitespace.
    - Disabling by default and hiding it under advanced options, no one will find it (except seasoned Emacs Helm users who have gotten accustomed to having this feature and will leave no stone unturned to find it if it's there).

- Keeping the search functionality as-is, since *Timelines* has fragment search, anyway, and its UI is better suited for this.
  - In that UI, it's clear the search applies per-message, since each node represents a message.
  - In the timeline view, it's possible to see all matching nodes at once.
  - The timeline view also highlights the matches in each node when you view that node. But in the main UI, there's a disconnect between viewing the list in *Manage chat files* and actually viewing a chat file. This in itself is good, but it complicates getting the highlight UX just right, if we want to have that.

### Priority

Low (Nice-to-have)

### Is this something you would be keen to implement?

Yes!

**Comment by Technologicat** [comment]
@Cohee1207: What is your opinion, would the *Manage Chat Files* view benefit from fragment search (swoop), or is it better if this functionality is only available in [*Timelines*](https://github.com/SillyTavern/SillyTavern-Timelines/)?

I could prepare a PR if this is desirable.

*[As usual, the obvious link is for the benefit of future readers coming in from Google.]*

**Comment by Cohee1207** [comment]
I don't have any preferences, the search function in Manage Chat Files is not very useful, since it doesn't show the search results (messages), only the files containing matches.

**Comment by Technologicat** [comment]
If it's ok, then I think I'll make a PR. :)

It's still somewhat useful in narrowing down the files that match a given topic.

As a first step, I could convert the search to a fragment search (as proposed above), with no other changes at this point. This would already be a small UX improvement over what we currently have.

Showing the matching messages requiers some UI/UX design. I think we could leave that part for later.

**Comment by Technologicat** [comment]
Made a minimal PR for this, posted as #1872.

**Comment by Technologicat** [comment]
Resolved by #1872.
